File: core/src/main/java/com/larryhsiao/nyx/core/jots/QueriedJots.java
package com.larryhsiao.nyx.core.jots;

import com.silverhetch.clotho.Source;
import org.locationtech.jts.geom.Point;
import org.locationtech.jts.io.WKTReader;

import java.sql.ResultSet;
import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import static java.lang.Double.MIN_VALUE;

/**
 * Adapter to adapt query result to Jot objects.
 */
public class QueriedJots implements Source<List<Jot>> {
    private final Source<ResultSet> query;

    public QueriedJots(Source<ResultSet> query) {
        this.query = query;
    }

    @Override
    public List<Jot> value() {
        try (ResultSet res = query.value()) {
            List<Jot> jots = new ArrayList<>();
            while (res.next()) {
                toJot(res, jots);
            }
            return jots;
        } catch (Exception e) {
            throw new IllegalArgumentException(e);
        }
    }

    private void toJot(ResultSet res, List<Jot> jots) throws Exception {
        Timestamp timestamp = res.getTimestamp(
            res.findColumn("createdTime"),
            Calendar.getInstance());
        String locationStr = res.getString("location");
        double[] location = new double[]{MIN_VALUE, MIN_VALUE};
        if (locationStr != null) {
            final Point locationGeo = new WKTReader().read(locationStr).getCentroid();
            location = new double[]{locationGeo.getX(), locationGeo.getY()};
        }
        jots.add(new ConstJot(
            res.getLong(res.findColumn("id")),
            res.getString(res.findColumn("content")),
            timestamp.getTime(),
            location,
            res.getString("mood"),
            res.getInt("version"),
            res.getInt("delete") == 1
        ));
    }
}

File: core/src/main/java/com/larryhsiao/nyx/core/jots/QueriedJots.kt
package com.larryhsiao.nyx.core.jots

import com.larryhsiao.clotho.Source
import org.locationtech.jts.io.WKTReader
import java.sql.ResultSet
import java.util.*

/**
 * Adapter to adapt query result to Jot objects.
 */
class QueriedJots(private val query: Source<ResultSet>) : Source<List<Jot?>?> {
    override fun value(): List<Jot> {
        try {
            query.value().use { res ->
                val jots: MutableList<Jot> = ArrayList()
                while (res.next()) {
                    toJot(res, jots)
                }
                return jots
            }
        } catch (e: Exception) {
            throw IllegalArgumentException(e)
        }
    }

    @Throws(Exception::class)
    private fun toJot(res: ResultSet, jots: MutableList<Jot>) {
        val timestamp = res.getTimestamp(
            res.findColumn("createdTime"),
            Calendar.getInstance())
        val locationStr = res.getString("location")
        var location = doubleArrayOf(Double.MIN_VALUE, Double.MIN_VALUE)
        if (locationStr != null) {
            val locationGeo = WKTReader().read(locationStr).centroid
            location = doubleArrayOf(locationGeo.x, locationGeo.y)
        }
        jots.add(ConstJot(
            res.getLong(res.findColumn("id")),
            res.getString(res.findColumn("title")),
            res.getString(res.findColumn("content")),
            timestamp.time,
            location,
            res.getString("mood"),
            res.getInt("version"),
            res.getInt("delete") == 1,
            res.getBoolean("private")
        ))
    }
}